### ISSUE TITLE
Babel Environment set to true breaks library configurations 

### ISSUE THREAD

**Issue by agneym** [webpro/knip#593]
Hey! Babel environment being set to `true` breaks configurations which are looking for a specific environment to add/remove configuration. 

https://github.com/webpro/knip/blob/c0c75e02ee58fbe365e479ca5a6db32cac478459/packages/knip/src/plugins/babel/helpers.ts#L35

Example scenario:

https://github.com/shakacode/shakapacker/blob/f902e186b102a45273d6dad4c6affb9da285763a/package/babel/preset.js#L5-L16

Can we use `process.env.NODE_ENV ?? process.env.BABEL_ENV` so as to get an accurate representation of the environment at that point in time?

**Comment by webpro** [comment]
By "breaks" you mean `currentEnv` is `true` instead of e.g. `development`, right?

Other than that, the idea seems sound to me: Knip will receive a configuration where all dependencies for all environments are included. How else could Knip figure this out in a single run? In general you don't want to run Knip multiple times I guess.

Having that said, and looking at [Babel source code](https://github.com/babel/babel/blob/main/packages/babel-core/src/config/helpers/config-api.ts), I think the function should be something like this:

```
{
  env: (env: string) => typeof env === 'string' ? true : 'development'
}
```

Would you agree?

**Comment by agneym** [comment]
@webpro yes, completely. I'm not entirely sure if the default should be 'development'. 

If we can use the environment variable that is actually passed to knip, the better. If not, 'development' should work.

**Comment by webpro** [comment]
That default value I got here: https://github.com/babel/babel/blob/main/packages/babel-core/src/config/helpers/environment.ts

The fix will come with the next release.

**Comment by webpro** [comment]
By the way I try to minimize reading things like `process.env` to maximize being runtime agnostic (e.g. run in Bun, Deno, whatnot).

**Comment by agneym** [comment]
@webpro Got it. Thank you 🙌

**Comment by agneym** [comment]
@webpro Now that I think about it, I think returning `true` for all environments might cause some babel configurations to be invalid. Developers use `isDevelopment`, `isProduction` kind of variables to configure plugins for babel. If we are returning just `true` for all environments, I would assume this causes at least some configurations to be invalid. 

I would expect something like [Babel's own implementation of the function](https://github.com/babel/babel/blob/33d8d7d54e1d2eb2d55c455a3ff34a73b55bd569/packages/babel-core/src/config/helpers/config-api.ts#L59-L73).

```
// Since we decided we aren't fetching from process
const defaultEnv = "development";

const env: EnvFunction = (
  value: string | string[] | (<T>(babelEnv: string) => T)
) => {
  if (typeof value === "undefined") return defaultEnv;
  if (typeof value === "function") {
    return value(defaultEnv);
  }
  return (Array.isArray(value) ? value : [value]).some((entry) => {
    if (typeof entry !== "string") {
      throw new Error("Unexpected non-string value");
    }
    return entry === defaultEnv;
  });
};

```

**Comment by webpro** [comment]
> Other than that, the idea seems sound to me: Knip will receive a configuration where all dependencies for all environments are included. How else could Knip figure this out in a single run? In general you don't want to run Knip multiple times I guess.

Yes in theory things might break, but so far so good, haven't received other issues around this.

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v5.10.0-canary.1. See [Release 5.10.0-canary.1](https://github.com/webpro/knip/releases/tag/5.10.0-canary.1) for release notes._

_Using Knip in a commercial project? Please consider [sponsoring me](https://github.com/sponsors/webpro)._
